satellite/overlay/selection_test.go

```diff
@@ -508,14 +540,37 @@ func TestFindStorageNodesDistinctNetworks(t *testing.T) {
 		require.NotEqual(t, nodes[0].LastIPPort, nodes[1].LastIPPort)
 		require.NotEqual(t, nodes[0].LastIPPort, excludedNodeAddr)
 		require.NotEqual(t, nodes[1].LastIPPort, excludedNodeAddr)
+		n2, err := satellite.Overlay.Service.SelectionCache.GetNodes(ctx, req)
+		require.NoError(t, err)
+		require.Len(t, n2, 2)
+		require.NotEqual(t, n2[0].LastIPPort, n2[1].LastIPPort)
+		require.NotEqual(t, n2[0].LastIPPort, excludedNodeAddr)
+		require.NotEqual(t, n2[1].LastIPPort, excludedNodeAddr)
+		n3, err := satellite.Overlay.Service.FindStorageNodesWithPreferences(ctx, req, &satellite.Config.Overlay.Node)
+		require.NoError(t, err)
+		require.Len(t, n3, 2)
+		require.NotEqual(t, n3[0].LastIPPort, n3[1].LastIPPort)
+		require.NotEqual(t, n3[0].LastIPPort, excludedNodeAddr)
+		require.NotEqual(t, n3[1].LastIPPort, excludedNodeAddr)
 
 		req = overlay.FindStorageNodesRequest{
-			MinimumRequiredNodes: 3,
-			RequestedCount:       3,
+			MinimumRequiredNodes: 4,
+			RequestedCount:       4,
 			ExcludedIDs:          excludedNodes,
 		}
-		_, err = satellite.Overlay.Service.FindStorageNodes(ctx, req)
+		n, err := satellite.Overlay.Service.FindStorageNodes(ctx, req)
+		require.Error(t, err)
+		n1, err := satellite.Overlay.Service.FindStorageNodesWithPreferences(ctx, req, &satellite.Config.Overlay.Node)
 		require.Error(t, err)
+		require.Equal(t, len(n), len(n1))
+		n2, err = satellite.Overlay.Service.SelectionCache.GetNodes(ctx, req)
+		require.Error(t, err)
+		// GetNodes returns 1 more node than FindStorageNodesWithPreferences because of the way the queries are...
+		// FindStorageNodesWithPreferences gets the IPs for the excludedNodeIDs and excludes all those IPs from the selection
+		// (which results in filtering out any node on the same network as a excludedNodeID),
+		// but the selection cache only filters IPs at time of selection which makes it so that it can include a node that shares a network
```

Comment:
hmm I'm not totally understanding why the selection cache only filters IPs at the time of selection/how it ends up with including a node that shares a network with an excluded ID. I guess this is a good quirk to be aware of. Is there a reason we wouldn't want GetNodes to return the same as FindStorageNodesWithPreferences?

Reply:
ideally the two methods return the same thing. However it doesn't make sense to me that FindStorageNodesWithPreferences filters out all nodes that share the same network with an excluded ID node so I wouldn't want to implement that into GetNodes.

Reply:
I think the original reason for that was to avoid a situation where a node is in the excluded ID list because we just uploaded pieces to it, and someone is running multiple nodes in the same network. In that case, we want to exclude that network so uploads are more evenly distributed.

Reply:
oh interesting...im curious when does that happen...like when do we upload pieces then try to upload more with excludedIDs?

Reply:
okay actually I think this case is unique to repairs where we just want to make sure we're not repairing pieces to nodes on an certain net if other nodes on that network are storing pieces from that segment. i.e. same kind of reason we're adding nodes with healthy pieces to the excluded nodes list in the repair process: https://github.com/storj/storj/blob/master/satellite/repair/repairer/segments.go#L174
does that make sense?

Reply:
that makes sense. 

So in that case sounds like there are many options to address this:
1) don't have repair use the selected nodes cache and instead just use the old way to select nodes
or
2) change the selected node cache to filter out all IPs that are in the excluded list
or 
3) leave everything the way it is and repair won't do that filter anymore

What do you think about those options?

I guess I am against 2 since its adds complexity to the selected node cache (due to repair) when really this optimization is only about uploads that the uplink does.  

I would bet 3 is fine, but hard to say exactly...im not sure if this is a legit risk about repairing to the same subnet. Not sure how often that would even occur, if at all, and if it does ever happen how often there would be negative consequences about it.

I guess then 1 seems like the best choice to me if it turns out to be really important that we filter out subnets of the excluded IDs, but I could be persuaded. Curious your perspective.

Reply:
I think 1 is the best choice/compromise for now.